app/models/llms/basic_workflow/prompting.py

```diff
@@ -0,0 +1,189 @@
+        Raises:
+            ValueError: If template is None or empty
+            TypeError: If values is not a dictionary
+        """
+        if template is None:
+            raise ValueError("Template cannot be None")
+        
+        if not isinstance(template, str):
+            raise TypeError(f"Template must be a string, got {type(template).__name__}")
+        
+        if not template.strip():
+            raise ValueError("Template cannot be empty or whitespace-only")
+        
+        if not isinstance(values, dict):
+            raise TypeError(f"Values must be a dictionary, got {type(values).__name__}")
+        
+        try:
+            return template.format(**values)
+        except KeyError as e:
+            missing_key = str(e).strip("'")
+            return self._format_with_fallbacks(template, values, missing_key)
+    
+    def _format_with_fallbacks(self, template: str, values: Dict[str, Any], missing_key: str) -> str:
+        """
+        Format template with fallback values for missing keys.
+        
+        Args:
+            template: The template string to format
+            values: Dictionary of available values
+            missing_key: The key that was missing from the values dictionary
+            
+        Returns:
+            str: Template formatted with fallbacks for missing keys
+        """
+        fallback_values = values.copy()
+
+        fallbacks = {
+            'city': 'the destination',
+            'days': 'several',
+            'month': 'your preferred time',
+            'preferences': 'your interests',
+            'budget': 'your budget',
+            'nation': 'the country',
+            'special_requirements': 'any special requirements',
+            'chat_history': 'previous conversation context',
+            'user_input': 'your travel request'
+        }
+        
+        if missing_key in fallbacks:
+            fallback_values[missing_key] = fallbacks[missing_key]
+        else:
+            fallback_values[missing_key] = f'[missing: {missing_key}]'
+        
+        try:
+            return template.format(**fallback_values)
+        except KeyError as e:
+            return self._replace_all_placeholders(template, fallback_values)
+    
+    def _replace_all_placeholders(self, template: str, values: Dict[str, Any]) -> str:
+        """
+        Replace all remaining placeholders with fallback values or generic placeholders.
+        
+        Args:
+            template: The template string with remaining placeholders
+            values: Dictionary of available values
+            
+        Returns:
+            str: Template with all placeholders replaced
+        """
+        import re
+        
+        placeholder_pattern = r'\{([^}]+)\}'
+        
+        def replace_placeholder(match):
+            key = match.group(1)
+            if key in values:
+                return str(values[key])
+            else:
+                return f'[placeholder: {key}]'
```

Comment:
replace_placeholder returns a string for every match, but if values contain non-string types, ensure they're serialized appropriately (e.g., dates, lists).